File: contrib/get_repository_tokens.py
#!/usr/bin/env python3

import requests
import re
import json
import sys
import getopt
 
 
s = requests.session()

def get_rels(response):
    links = response["link"].split(',')
    regex = re.compile(r'<(.*?)>; rel="(\w+)"')
    rels_hash = {}
    for link in links:
        href, name = regex.findall(link)[0]
        rels_hash[name] = href
    return rels_hash
                        

def get_token(match_repo_url):
    resp = s.get("https://scc.suse.com/connect/organizations/repositories", headers={'Accept':'application/vnd.scc.suse.com.v4+json','cache-control':'no-cache'})
    repositories = []

    while True:
        for repository in resp.json():
            if match_repo_url in repository['url'].split("?")[0]:
                print repository['url'].split("?")[0]+","+repository['url'].split("?")[-1]
        rels = get_rels(resp.headers)
        if not 'next' in rels:
            break
        resp = s.get(rels['next'])

def print_USAGE():
    print 'Usage:'
    print '\t'+self+' -u <username> -o <password> <part of repository URL> [... <part of repository URL>]'
    print '\tUse the SCC mirror credentials of your organization for user and password'
    print '\tProvide strings to match against repository URL'
    print '\tThe script provides all matches, one per line, full URL and TOKEN separated by ",".'

def main(argv):
    user = ''
    password = ''
    try:
        opts, args = getopt.getopt(argv,"u:p:h",["user=","password=","help"])
    except getopt.GetoptError:
        print_USAGE()
        sys.exit(2)
    for opt, arg in opts:
        if opt in ('-h','--help'):
            print_USAGE()
            sys.exit()
        elif opt in ("-u", "--user"):
            user = arg
        elif opt in ("-p", "--password"):
            password = arg
    if not user or not password:
        print 'ERROR: missing user or password of your mirror credentials'
        print_USAGE()
        sys.exit()
    if not args:
        print 'ERROR: no strings to match against repository URL'
        print_USAGE()
        sys.exit()

    s.auth = (user, password)
    for match_repo_url in args:
        get_token(match_repo_url)
 

if __name__ == "__main__":

    self= sys.argv[0]
    main(sys.argv[1:])

File: contrib/get_repository_tokens.rb
#!/usr/bin/env ruby
require 'net/http'
require 'optparse'
require 'json'

def process_rels(headers)
  links = (headers['link'].first || '').split(', ').map do |link|
    href, name = link.match(/<(.*?)>; rel="(\w+)"/).captures
    [name.to_sym, href]
  end
  Hash[*links.flatten]
end

def get_token(repo_url)
  url = 'https://scc.suse.com/connect/organizations/repositories'
  response = get(url)

  loop do
    headers = response.header
    data = JSON.parse response.body
    data.each do |repo|
      next unless repo['url'].include? repo_url

      puts('Found a match')
      puts('URL: ' + repo['url'].split('?').first + "\n" + 'Token: ' + repo['url'].split('?').last)
    end
    rels = process_rels(headers.to_hash)
    unless rels.key? :next
      break
    end

    response = get(rels[:next])
  end
end

def get(url)
  uri = URI(url)

  request = Net::HTTP::Get.new uri
  request.basic_auth(@username, @password)

  request['cache-control'] = 'no-cache'
  request['Accept'] = 'application/vnd.scc.suse.com.v4+json'

  response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
    http.request(request)
  end

  response
end

@username = ''
@password = ''

def main
  OptionParser.new do |opts|
    opts.banner += ' [Repositories]'
    opts.separator 'Get repository CDN tokens'
    opts.version = '0.0.1'

    opts.on('-u', '--user USERNAME',
            '',
            'Mirroring username') { |arg| @username = arg }

    opts.on('-p', '--password PASSWORD',
            '',
            'Mirroring password') { |arg| @password = arg }
    begin
      opts.parse!
    rescue OptionParser::ParseError => e
      warn e
      warn '(-h or --help will show valid options)'
      exit 1
    end
  end

  if ARGV.empty?
    warn "Please add the repository you're searching for"
    exit 1
  else
    ARGV.each do |match_repo_url|
      get_token(match_repo_url)
    end
  end
end

main
